source/octf/analytics/statistics/Distribution.cpp

```diff
@@ -0,0 +1,176 @@
+            : Sum(size)
+            , Count(size) {
+        Id = id;
+        Size = size;
+        Power = power;
+        Begin = 0;
+        if (id) {
+            Begin = std::pow(power, id - 1) * size;
+        }
+        End = std::pow(power, id) * size - 1;
+        Step = (End - Begin + 1) / size;
+    }
+
+    void operator+=(uint64_t value) {
+        if (value < Begin || value > End) {
+            throw Exception("Updating wrong distribution bucket");
+        }
+
+        uint id = value;
+        id -= Begin;
+        id /= Step;
+
+        if (id < Size) {
+            Sum[id] += value;
+            Count[id]++;
+        } else {
+            throw Exception("Updating distribution bucket fatal error");
+        }
+    }
+
+    uint64_t Id;
+    uint64_t Size;
+    uint64_t Power;
+    uint64_t Begin;
+    uint64_t End;
+    uint64_t Step;
+
+    std::vector<uint64_t> Sum;
+    std::vector<uint64_t> Count;
+};
+
+Distribution::Distribution(const std::string &unit,
+                           uint64_t bucketSize,
+                           uint64_t backetPower)
+        : m_unit(unit)
+        , m_bucketSize(bucketSize)
+        , m_bucketPower(backetPower)
+        , m_total()
+        , m_count()
+        , m_min(std::numeric_limits<decltype(m_min)>::max())
+        , m_max()
+        , m_histogram() {
+    if (backetPower < 2) {
+        throw Exception("Invalid bucket power");
+    }
+}
+
+Distribution::Distribution(const Distribution &other)
+        : m_unit(other.m_unit)
+        , m_bucketSize(other.m_bucketSize)
+        , m_bucketPower(other.m_bucketPower)
+        , m_total(other.m_total)
+        , m_count(other.m_count)
+        , m_min(other.m_min)
+        , m_max(other.m_max)
+        , m_histogram(other.m_histogram) {}
+
+Distribution &Distribution::operator=(const Distribution &other) {
+    if (this != &other) {
+        m_unit = other.m_unit;
+        m_bucketSize = other.m_bucketSize;
+        m_bucketPower = other.m_bucketPower;
+        m_total = other.m_total;
+        m_count = other.m_count;
+        m_min = other.m_min;
+        m_max = other.m_max;
+        m_histogram = other.m_histogram;
+    }
+
+    return *this;
+}
+
+Distribution::~Distribution() {}
+
+void Distribution::operator+=(uint64_t value) {
+    m_total += value;
+    m_count++;
+    m_min = std::min(m_min, value);
+    m_max = std::max(m_max, value);
+
+    // Update histogram
+    auto &backet = getBucket(value);
+    backet += value;
+}
+
+void Distribution::fillDistribution(proto::Distribution *distribution) const {
+    distribution->Clear();
+    distribution->set_avarage(m_count ? m_total / m_count : 0);
+    distribution->set_min(m_count ? m_min : 0);
+    distribution->set_max(m_max);
+    distribution->set_count(m_count);
+    distribution->set_total(m_total);
+    distribution->set_unit(m_unit);
+}
+
+Distribution::Bucket &Distribution::getBucket(uint64_t value) {
+    uint64_t i = 0;
```

Comment:
Is i the bucket level? If so please rename 'i' to 'level' and start from 1 - for clarity and consistency with header docs